### PR TITLE
[TESTERS NEEDED!] Savestates: Compressed files

### PR DESCRIPTION
**Pull request by elad335**
Along with #14784, this pull requests aims to reduce significantly the file size of savestates which can anywhere from few hundreds of MBs to few GBs or even more.
Initial measurements suggest more size reduction of more than 50 percent! But it is probably more in most cases, need user testing!

TODO:

- [x] Make it compatible with TAR serialization.
- [x] Make TAR loader RAM efficient as well, it was found that the previous code loads entire files on RAM when reading TAR entries, so this will get fixed here.
- [x] Test it.
- [ ] #14784 also allows for asynchronous savestate file writes when I think about it, this would would increase the speed of saving and is now targeted for this pr. Note: postponed to another pull request.

Additional bugfixes:
- [x] Fix missing TAR last entry in savestate in some cases. 
- [x] Fix a bug in saved HLE state in sys_lwcond/sys_lwmutex/sys_mutex/sys_cond syscalls.

### CODE REVIEW

File: rpcs3/util/serialization.hpp [hunk 1]
```diff
@@ -40,7 +40,16 @@ namespace utils
...
 
 		virtual bool handle_file_op(serial& ar, usz pos, usz size, const void* data = nullptr) = 0;
-		virtual usz get_size(const utils::serial& ar, usz recommended = umax) const = 0;
+
+		virtual usz get_size(const utils::serial& ar, usz recommended = umax) const
+		{
+			return 0;
+		}
+
+		virtual bool is_null()
```
Comment (Megamouse) [inline rpcs3/util/serialization.hpp:49]: const ?

**Comment by mirh** [comment]
I don't want to start a codec war, but certainly in this day and age there are better alternatives than zlib?
(yes, including also for performance if that was the most paramount concern)

**Comment by elad335** [comment]
Okay. I only have one requirement for them: it needs to be able to compress data chunks with the ability to perform well even if not all data is known yet.

**Comment by mirh** [comment]
I'm going out on a limb, and guess that just like in every other case ever you can check xz/lzma2 for usually the best compression, and then zstd or brotli for some slight compromise with speed.

File: rpcs3/Emu/RSX/RSXThread.cpp [hunk 2]
```diff
@@ -3594,36 +3597,34 @@ namespace rsx
...
+				save_manager.m_file_handler = make_compressed_serialization_file_handler(temp.file);
+				save_manager(frame_capture);
 
-			fs::pending_file temp(file_path);
+				save_manager.breathe(true);
 
-			if (temp.file && (temp.file.write(save_manager.data), temp.commit(false)))
-			{
-				rsx_log.success("Capture successful: %s", file_path);
+				if (!temp.commit(false))
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/RSXThread.cpp:3613]: shouldn't this be ``if (temp.commit(false))``

**Comment by elad335** [comment]
56% compression rate and the file is valid, still RPCS3 won't load it but it is work in progress.
image

**Comment by elad335** [comment]
Applied some bugfixes, loading of files is also starting to work!

File: rpcs3/Emu/System.cpp [hunk 3]
```diff
@@ -2309,33 +2349,87 @@ void Emulator::FixGuestTime()
 
 void Emulator::FinalizeRunRequest()
 {
-	auto on_select = [](u32, spu_thread& spu)
+	const bool autostart = !!g_cfg.misc.autostart;
+
+	static const auto cpu_op = [](atomic_bs_t<cpu_flag>& state, bs_t<cpu_flag> add, bs_t<cpu_flag> sub)
```
Comment (Megamouse) [inline rpcs3/Emu/System.cpp:2354]: these functions really don't belong here

File: rpcs3/Emu/System.cpp [hunk 4]
```diff
@@ -896,13 +896,31 @@ game_boot_result Emulator::Load(const std::string& title_id, bool is_disc_patch,
 		}
 		else
 		{
-			if (fs::file save{m_path, fs::isfile + fs::read}; save && save.size() >= 8 && save.read<u64>() == "RPCS3SAV"_u64)
+			fs::file save{m_path, fs::isfile + fs::read};
+
+			if (!m_path.ends_with(".gz") && save && save.size() >= 8 && save.read<u64>() == "RPCS3SAV"_u64)
```
Comment (Megamouse) [inline rpcs3/Emu/System.cpp:901]: Tbh, I think this is getting out of hand. Every section of code that has `m_ar->` in it belongs to another file.
Half of System.cpp is just writing or reading random stuff to that object.
I'd even be happy if it's just a second cpp with Emulator internal functions that only deal with the savestate.

**Comment by Ordinary205** [comment]
This may sound like a silly question, but do I have to test this PR just to see if the files has significantly reduced? Or do I need to test this PR just to see if there's any savestate errors on other games? Because Demon's Souls seems to behave fine after loading.
Demon's Souls
I've tested the performance on offical and PR build from each comparison, and haven't noticed a slight regression.
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/13472021/RPCS3.log.gz)

**Comment by elad335** [comment]
The performance measurement needed here is for saving time itself, and also test file size by comparing file size to the one after extraction using 7zip or WinRAR etc. Although I know there is a a bug here preventing some games to reload properly such as Skate 3 or TLoU.

**Comment by elad335** [comment]
Scratch that, I found the bug!! Will merge and you are welcome to test on master!

**Pull request merged**